In [6]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix

In [7]:
import sys
import os

sys.path.append(os.path.abspath(".."))

In [8]:
df=pd.read_csv("../data/raw/credit_dataset.csv")

X = df[[
    "age",
    "income",
    "debt_to_income",
    "credit_score",
    "loan_amount"
]]

y = df["approved"]

In [9]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.3,
    random_state=42,
    stratify=y
)

In [10]:
scaler = StandardScaler()

X_train_scaled=scaler.fit_transform(X_train)
X_test_scaled=scaler.fit_transform(X_test)

In [11]:
knn=KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train_scaled, y_train)

KNeighborsClassifier()

In [12]:
y_pred_knn=knn.predict(X_test_scaled)

In [13]:
cm_knn=confusion_matrix(y_test,y_pred_knn)
cm_knn

array([[ 987,  213],
       [ 163, 1637]])

In [14]:
cost_fp = 100
cost_fn = 10

TN, FP, FN, TP = cm_knn.ravel()

expected_cost_knn = FP * cost_fp + FN * cost_fn
cost_per_applicant_knn = expected_cost_knn / len(y_test)

expected_cost_knn, cost_per_applicant_knn

(22930, 7.6433333333333335)

1. Logistic Regression matrix ==
[[ 413  787]
 [ 197 1603]]
80670, 26.89

KNN matrix ==
([[ 987,  213],
 [ 163, 1637]])
22930, 7.643

Therefore overall the FP and FP are both reduced.

2. Expected cost is lower than before.

3. Based on cost alone KNN is better than Logistic Regression.

4. KNN seems to be better suited for this use case. Specifically since it is reducing the cost of errors

5. Not deploy. The model is still not optimised for reducing FP's. It is only better at scoring. 
